kernel/src/process_load_utilities.rs

```diff
@@ -0,0 +1,791 @@
+                    
+                    //we return sram_end_addresses
+                    let addresses = process.get_addresses();
+                    sram_end_addresses = addresses.sram_end;
+
+                    //we return process_copy
+                    process_copy = Some(process);
+                });
+            }
+            else {
+                //header length 0 means invalid header
+                return Err(ProcessLoadError::InternalError);
+            }
+        }
+
+        Ok((sram_end_addresses, process_copy))
+    }
+
+    // In order to match the result value of command
+    fn load_processes_air(
+        &self,
+        proc_data: &mut ProcLoaderData,
+    ) -> Result<(), ErrorCode> {
+        
+        let res = self.load_processes_advanced_air(proc_data);
+
+        // Without alignment error, we only store the entry point, the start address, and the size of the flashed application
+        match res{
+            Ok((sram_end, process_copy)) => {
+                // This variable will be used, when loading the another app at next load work by ota app
+                // This is necessary to prevent the access violation of sram memory whilch are already used by kernel and other apps.
+                proc_data.dynamic_unsued_sram_start_addr = sram_end;
+                
+                // Store the entry point, the start address, and the size of the flashed application into PROCESS global array
+                // Although I used unsafe keyword, I think it's okay, becasue we pass the exact pointer of PROCESS global array
+                unsafe {
+                    *self.ptr_process.offset(proc_data.index.try_into().unwrap()) = process_copy;
+
+                    // We also save process region information to check the validity of 'proc_data.dynamic_flash_start_addr' in future load work
+                    *self.ptr_process_region_start_address.offset(proc_data.index.try_into().unwrap()) = proc_data.dynamic_flash_start_addr;
+                    *self.ptr_process_region_size.offset(proc_data.index.try_into().unwrap()) = proc_data.appsize_requested_by_ota_app;
+                }
+
+                // We increase the index for next load work by OTA app
+                proc_data.index += 1;
+
+                return Ok(());
+            }
+            Err(_e) => {
+                // If there is an error caused by misalignment,
+                // 'proc_data.dynamic_unsued_sram_start_addr' and 'proc_data.index' will hold current unused sram start address
+                return Err(ErrorCode::FAIL);
+            }
+        }
+    }
+    
+    // This function is implemented based on load_processes_advanced
+    // the purpose is to parse the dynamically changing start address of flash memory satisfying MPU rules
+    fn find_dynamic_start_address_of_writable_flash_advanced(
+        &self,
+        proc_data: &mut ProcLoaderData,
+        start_app: usize,
+    ) -> Result<(), ProcessLoadError> {
```

Comment:
Implementing this by scanning flash makes me nervous, one because it duplicates TBF parsing code, and two because it uses `unsafe`. Did you try implementing this by scanning the `PROCESSES` array, and using the already flash region addresses and sizes contained in the `Process` structs instead?

Reply:
Yes. I can handle it. I totally agree with your feedback! I need to improve my code in terms of safety!